raft: stop resetting electionElapsed when attempting to campaign

iskettaneh opened the pull request:
This commit changes the current behaviour where we reset electionElapsed when campaigning (to allow us to campaign again on the next randomizedElectionTimeout). This problem with the current behaviour is that if a follower can't campaign (due to store liveness limitations), resetting the electionElapsed will re-arm the inHeartbeatLease, making the follower reject MsgVotes.

This commit changes that by campaigning at every randomizedElectionTimeout interval.

Epic: None

Release note: None

iskettaneh committed:
raft: stop resetting electionElapsed when attempting to campaign

This commit changes the current behaviour where we reset electionElapsed
when campaigning (to allow us to campaign again on the next
randomizedElectionTimeout). This problem with the current behaviour is
that it could prevent a preCandidate from voting to other candidates
because it will be in the heartbeat lease.

This commit changes that by campaigning at every
randomizedElectionTimeout interval.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/133418)
<!-- Reviewable:end -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 1128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OA04JmO82d670ymR_Gb:-OA04JmO82d670ymR_Gc:b6ktb9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/353fa4043b1431eb8268604a898f4ccef398d5a9/pkg/raft/raft.go#L1128)):*
> ```Go
> 	}
> 
> 	if r.atRandomizedElectionTimeout() {
> ```

@arulajmani @nvanbenschoten I just realized that this changes one current behaviour:

Previously, if we are not promotable, we would check again at every tick after the RandomizedElectionTimeout. However, with this change, we will check again on the next RandomizedElectionTimeout.

I could check if we are not promotable, and subtract one from the elapsedElection, but I just want to see what you think and whether you think I should do that or not


<!-- Sent from Reviewable.io -->

nvanbenschoten commented on pkg/raft/raft.go:
```diff
@@ -398,7 +398,7 @@ type raft struct {
 	//
 	// Invariant: electionElapsed = 0 when r.leadEpoch != 0 on a follower.
-	electionElapsed int
+	electionElapsed uint64
```
We're performing arithmetic on these fields, so I think we should make them `int64`s, not `uint64`s. Unsigned integers are generally frowned upon as a means of encoding "non-negative semantics". See https://graphitemaster.github.io/aau/ and https://www.learncpp.com/cpp-tutorial/unsigned-integers-and-why-to-avoid-them for some discussion why. I recall there was also a similar best practice at Google. They don't actually enforce that values don't go negative, they just underflow and cause such bugs to be harder to catch. It's also common to want to use a negative offset when manipulating values that should not go negative.

nvanbenschoten commented on pkg/base/config.go:
```diff
@@ -524,15 +524,15 @@ type RaftConfig struct {
 	// RaftReproposalTimeoutTicks is the number of ticks before reproposing a Raft
 	// command. This also specifies the number of ticks between each reproposal
 	// check, so the actual timeout is 1-2 times this value.
 	RaftReproposalTimeoutTicks int
```
If we're going to change the data type for "tick" units, let's do so everywhere.

nvanbenschoten commented on pkg/base/config.go:
```diff
@@ -633,10 +633,10 @@ func (cfg *RaftConfig) SetDefaults() {
 	}
 	if cfg.RaftElectionTimeoutTicks == 0 {
-		cfg.RaftElectionTimeoutTicks = defaultRaftElectionTimeoutTicks
+		cfg.RaftElectionTimeoutTicks = uint64(defaultRaftElectionTimeoutTicks)
```
Can we change the type of `defaultRaftElectionTimeoutTicks` and `defaultRaftHeartbeatIntervalTicks`? `EnvOrDefaultInt64` exists, if you agree with my other comment.

iskettaneh committed:
raft: make election and heartbeat ticks of type int64

In the near future, we want to keep incrementing the electionElapsed
until we hear from a leader. In order to avoid any chances of an integer
overflow, this commit changes the electionElapsed (along with other
symmetrical variables from int to int64)

Epic: None

Release note: None

iskettaneh committed:
raft: stop treating promotable() specially when campaigning

Right now, there are multiple reasons why a replica can't campaign.
However, we treat the !r.promotable() differently. If a replica is not
promotable and we are past the randomizedElectionTimeout, we will keep
try to campaign on every tick. This is different from the other cases
inside hup() where we discover that we can't campaign. In those cases,
we attempt to campaign after a RandomizedElectionTimeout period.

It was introduced in this commit: 7b53c4291fd1 which made followers
ignore MsgVote if they recently heard from the leader. The way the code
was structured, it used to reset the electionElapsed on every
electionTick if it wasn't promotable, meaning that the timer will start
from the moment the replica is promotable. However, with the
introduction of a heartbeat lease. That would have meant that a non
promotable replica is always in a heartbeat lease and can't vote in an
election. Therefore, the change seems to have changed the behaviour a
bit and only reset the electionElapsed if both we are past the election
timeout and we are promotable.

This change reverts the old behaviour where we wait an entire
RandomizedElectionTimeout to campaign if we are not promotable.

To be on the safer side, we will monitor this change on master first
before backporting it to 24.3

Epic: None

Release note: None

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/base/config.go` line 532 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133418#gh-1816955050:-OA4YU2X-TiZrgSJZfOe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a653dc82c9d4b84582ad5241a9b0eff31638320e/pkg/base/config.go#L532)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If we're going to change the data type for "tick" units, let's do so everywhere.
</blockquote></details>

Done.
___
*[`pkg/base/config.go` line 636 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133418#gh-1816956336:-OA4YV0dDHjUxrtkgS-D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a653dc82c9d4b84582ad5241a9b0eff31638320e/pkg/base/config.go#L636)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we change the type of `defaultRaftElectionTimeoutTicks` and `defaultRaftHeartbeatIntervalTicks`? `EnvOrDefaultInt64` exists, if you agree with my other comment.
</blockquote></details>

Done.
___
*[`pkg/raft/raft.go` line 401 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133418#gh-1816952127:-OA4D-u0B50ykZpoGrDN:b-epa4a9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a653dc82c9d4b84582ad5241a9b0eff31638320e/pkg/raft/raft.go#L401)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We're performing arithmetic on these fields, so I think we should make them `int64`s, not `uint64`s. Unsigned integers are generally frowned upon as a means of encoding "non-negative semantics". See https://graphitemaster.github.io/aau/ and https://www.learncpp.com/cpp-tutorial/unsigned-integers-and-why-to-avoid-them for some discussion why. I recall there was also a similar best practice at Google. They don't actually enforce that values don't go negative, they just underflow and cause such bugs to be harder to catch. It's also common to want to use a negative offset when manipulating values that should not go negative.
</blockquote></details>

Yeah I had the same concern yesterday after I pushed this PR. I will change it to int64 to be on the safer side


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 19 of 19 files at r4, 3 of 3 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @iskettaneh)

___
*[`-- commits` line 7 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJDY5V2y89aNDfW8v9):*
"int64"
___
*[`-- commits` line 28 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJHdO7BuuKYVjdth06):*
"introduction of"
___
*[`pkg/raft/raft.go` line 83 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJHAcu9coCb0tH5RyG:-OAJHAcu9coCb0tH5RyH:bpofp1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/73f7b3999ae7fa5fc02cb70a99d4e759a833c8b1/pkg/raft/raft.go#L83)):*
> ```Go
> }
> 
> func (r *lockedRand) Intn64(n int64) int64 {
> ```

The `math/rand` package calls this `Int63n` (https://pkg.go.dev/math/rand#Rand.Int63n), so we should do the same here.
___
*[`pkg/raft/raft.go` line 395 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJIm0d-JPTH0xGidYJ:-OAJIm0d-JPTH0xGidYK:bx1roza) ([raw file](https://github.com/cockroachdb/cockroach/blob/90dac76795db8f32b1d983b6c45446f407ffdc7b/pkg/raft/raft.go#L395)):*
> ```Go
> 	uncommittedSize entryPayloadSize
> 
> 	// electionElapsed is the number of ticks since we last reached the
> ```

We will need to update this comment.


<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 83 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJHAcu9coCb0tH5RyG:-OAJMAsX6g6MOKjbwNU1:b-896fja) ([raw file](https://github.com/cockroachdb/cockroach/blob/73f7b3999ae7fa5fc02cb70a99d4e759a833c8b1/pkg/raft/raft.go#L83)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The `math/rand` package calls this `Int63n` (https://pkg.go.dev/math/rand#Rand.Int63n), so we should do the same here.
</blockquote></details>

Done!


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 2 of 3 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/raft/raft.go` line 395 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJIm0d-JPTH0xGidYJ:-OAJRuhp9EMs2rMDQgj_:b-phtd5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/90dac76795db8f32b1d983b6c45446f407ffdc7b/pkg/raft/raft.go#L395)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We will need to update this comment.
</blockquote></details>

"or if the leader is fortified when ticked"

I've read this part (which was there before) a few times and don't understand it. Do you?


<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/raft/raft.go` line 395 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJIm0d-JPTH0xGidYJ:-OAJSNUHD03crZDzdwu6:b5epgdn) ([raw file](https://github.com/cockroachdb/cockroach/blob/90dac76795db8f32b1d983b6c45446f407ffdc7b/pkg/raft/raft.go#L395)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"or if the leader is fortified when ticked"

I've read this part (which was there before) a few times and don't understand it. Do you?
</blockquote></details>

My understanding is this: for a follower, as long as it's fortifying a leader, the elapsedElection is zero.
Basically, elapsedElection is set to zero when are fortifying a leader, and we don't increment the elapsedElection if we are fortifying a leader.

I think a simpler sentence would be: "elapsedElection is set to 0 for followers who are fortifying a leader". WDYT?


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/raft/raft.go` line 395 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJIm0d-JPTH0xGidYJ:-OAJbPO-1GzU9P5qbEtm:boodzdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/90dac76795db8f32b1d983b6c45446f407ffdc7b/pkg/raft/raft.go#L395)):*
> for a follower, as long as it's fortifying a leader, the elapsedElection is zero.

Agreed, I just don't understand the sentence. Maybe I'm reading it wrong. What you suggest sounds good (and is captured in the Invariant mentioned below), although it might be nice to not have two sentences which directly contradict each other back-to-back. 

How about "For followers, it is the number of ticks since they last received a valid message from the from the current leader, unless the follower is fortifying a leader (`leadEpoch != 0`), in which case it is always set to 0."


<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/raft/raft.go` line 395 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJIm0d-JPTH0xGidYJ:-OAJtS7-5o3UQ0tQ4ftP:b-896fja) ([raw file](https://github.com/cockroachdb/cockroach/blob/90dac76795db8f32b1d983b6c45446f407ffdc7b/pkg/raft/raft.go#L395)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> for a follower, as long as it's fortifying a leader, the elapsedElection is zero.

Agreed, I just don't understand the sentence. Maybe I'm reading it wrong. What you suggest sounds good (and is captured in the Invariant mentioned below), although it might be nice to not have two sentences which directly contradict each other back-to-back. 

How about "For followers, it is the number of ticks since they last received a valid message from the from the current leader, unless the follower is fortifying a leader (`leadEpoch != 0`), in which case it is always set to 0."
</blockquote></details>

Done!


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-:-OAK5X6ZE2p9sH0SejUl:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani)



<!-- Sent from Reviewable.io -->

iskettaneh commented:
TFTR!

bors r+

iskettaneh commented:
bors r+

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -1184,7 +1184,7 @@ func (r *raft) tickElection() {
 		r.electionElapsed++
 	}
 
-	if r.promotable() && r.pastElectionTimeout() {
```
From you second commit description: 
> If a replica is not promotable and we are past the randomizedElectionTimeout, we will keep try to campaign on every tick. 

I’m not following why we do that. The `promotable()` check ensures 3 things:
1. we are part of the active configuration. If not we don’t have a chance to win - nodoby will vote for us
2. we are not a learner. If we are, we aren’t a voter and shouldn’t campaign
3. we aren’t processing an incoming snapshot. I’m less clear on why we can’t campaign if we are

At least in cases 1 and 2, why would we proceed to campaign here?

Are we relying on the fact that `hup` method will check `promotable()` down the stack and not campaign anyway?

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -2647,11 +2647,11 @@ func (r *raft) loadState(state pb.HardState) {
+// atRandomizedElectionTimeout returns true if r.electionElapsed modulo the
+// r.randomizedElectionTimeout is equal to 0. This means that at every
+// r.randomizedElectionTimeout period, this method will return true once.
+func (r *raft) atRandomizedElectionTimeout() bool {
```
The caller of this doesn't check `promotable()` anymore, so what happens in the following case?

1. tick `randomizedElectionTimeout`: we call `hup`, but `promotable()` is false, so we don't campaign
2. tick `randomizedElectionTimeout+1`: we became promotable, but we will have to wait out another `randomizedElectionTimeout` to campaign

Is this intended?
